### PR TITLE
[Transform] Remove duplicate checkpoint audits

### LINKED ISSUE #105106: TransformIndexer might write duplicate audit messages
### Elasticsearch Version

-

### Installed Plugins

_No response_

### Java Version

_bundled_

### OS Version

-

### Problem Description

`logCount` in the `TransformIndexer` is `volatile` => could be updated by multiple threads?
But the `++` increment operation is not atomic:
https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/TransformIndexer.java#L1167

It could happen that some increments will be lost / duplicate audit messages will be written.

CC @hendrikmuhs 

### Steps to Reproduce

-

### Logs (if relevant)

_No response_

### PR DESCRIPTION
**Pull request by prwhelan**
Transform Checkpoints have a chance to log duplicate audits or drop iterations.  The volatile counters can be read and incremented in multiple threads, potentially storing the same value back into memory.

Replacing volatile counters with a single Atomic counter, which counts down the iterations until it reaches zero, then updates the counter to the next audited checkpoint.

Closes #105106

### CODE REVIEW

File: x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/TransformIndexer.java [hunk 1]
```diff
@@ -1155,25 +1155,31 @@ private SearchRequest buildQueryToUpdateDestinationIndex() {
 
     /**
      * Indicates if an audit message should be written when onFinish is called for the given checkpoint
-     * We audit the first checkpoint, and then every 10 checkpoints until completedCheckpoint == 99
-     * Then we audit every 100, until completedCheckpoint == 999
+     * We audit every checkpoint for the first 10 checkpoints until completedCheckpoint = 9.
```
Comment (przemekwitek) [inline x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/TransformIndexer.java:1158]: Please change to "until completedCheckpoint == 9" for consistency with below statements.

File: x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/TransformIndexer.java [hunk 2]
```diff
@@ -1155,25 +1155,31 @@ private SearchRequest buildQueryToUpdateDestinationIndex() {
...
+     * Then we audit every 10 checkpoints until completedCheckpoint == 99.
```
Comment (przemekwitek) [inline x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/TransformIndexer.java:1159]: Sorry, I'm not a native speaker so you will be better to judge here:
Could it be written "every 10th checkpoint", "every 100th checkpoint" etc.?

File: x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/TransformIndexer.java [hunk 3]
```diff
@@ -1155,25 +1155,31 @@ private SearchRequest buildQueryToUpdateDestinationIndex() {
...
-            return false;
-        }
-        if (completedCheckpoint == 0) {
-            return true;
-        }
-        int log10Checkpoint = (int) Math.floor(Math.log10(completedCheckpoint));
-        logEvery = log10Checkpoint >= 3 ? 1_000 : (int) Math.pow(10.0, log10Checkpoint);
-        logCount = 0;
-        return true;
+        return remainingCheckpointsToSkipAuditing.getAndUpdate(count -> {
```
Comment (przemekwitek) [inline x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/TransformIndexer.java:1168]: Nice idea!
This way we achieve atomicity of the update operation but also improve performance as we do not perform heavy mathematical operations.

File: x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/TransformIndexer.java (hunk 2 above)
Comment (prwhelan) [reply x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/TransformIndexer.java:1159]: Yes, that does sound a bit clearer. "Then we audit every 10th checkpoint until completedCheckpoint == 99."

File: x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/TransformIndexer.java (hunk 3 above)
Comment (prwhelan) [reply x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/TransformIndexer.java:1168]: Thank you!

File: x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/TransformIndexer.java [hunk 4]
```diff
@@ -119,8 +120,7 @@ private enum RunState {
     private Map<String, Object> nextChangeCollectorBucketPosition = null;
 
     private volatile Integer initialConfiguredPageSize;
-    private volatile long logEvery = 1;
-    private volatile long logCount = 0;
+    private final AtomicInteger remainingCheckpointsToSkipAuditing = new AtomicInteger(0);
```
Comment (sabi0) [inline x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/TransformIndexer.java:123]: What do you think of `remainingCheckpointsUntilAudit` ?

File: x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/TransformIndexer.java [hunk 5]
```diff
@@ -1155,25 +1155,31 @@ private SearchRequest buildQueryToUpdateDestinationIndex() {
...
     /**
      * Indicates if an audit message should be written when onFinish is called for the given checkpoint
-     * We audit the first checkpoint, and then every 10 checkpoints until completedCheckpoint == 99
-     * Then we audit every 100, until completedCheckpoint == 999
+     * We audit every checkpoint for the first 10 checkpoints until completedCheckpoint = 9.
+     * Then we audit every 10 checkpoints until completedCheckpoint == 99.
+     * Then we audit every 100, until completedCheckpoint == 999.
      *
-     * Then we always audit every 1_000 checkpoints
+     * Then we always audit every 1_000 checkpoints.
```
Comment (sabi0) [inline x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/TransformIndexer.java:1162]: Cosmetic: I think this line could be joined to the paragraph above.

**Review by przemekwitek — APPROVED** [review]
LGTM

**Pull request merged**